pkg/server/server.go

```diff
@@ -1,129 +1,158 @@
+
+		// send data if we received any
+		if br > 0 {
+			conf.log.WithField("session", reqId).Debugf("writing %d bytes to conn", br)
+			_, err := session.Conn.Write(data)
+			if err != nil {
+				conf.log.WithError(err).WithField("session", reqId).Errorf("failed writing data to socket")
+				message.ShouldClose = true
 			} else {
-				data := message.GetData()
-				if len(data) > 0 {
-					conn := session.Conn
-					_, err := conn.Write(data)
-					if err != nil {
-						log.Errorf("%s; failed writing data to socket", reqId)
-					}
-				}
-				if message.ShouldClose == true {
-					ok, _ := common.CloseSession(reqId)
-					if ok != true {
-						log.Errorf("%s; failed closing session", reqId)
-					}
-				}
+				conf.log.WithField("session", reqId).Debugf("wrote %d bytes to conn", br)
 			}
 		}
+
+		if message.ShouldClose == true {
+			conf.log.WithField("session", reqId).Debug("closing session")
+			session.Close()
+			conf.log.WithField("session", reqId).Debug("closed session")
+		}
 	}
 }
 
-func readConn(session *common.Session, sessions chan<- *common.Session) {
-	log.Debugf("got new session %s", session.Id.String())
+func readConn(conf *ServerConfig, session *common.Session, sessions chan<- *common.Session) {
+	conf.log.WithField("session", session.Id.String()).Info("new conn connection")
 
-	sessions <- session
-	// We want to inform the client that we accepted a connection - some weird ass protocols wait for data from the server when connecting
-	// Read from socket in a loop and push messages to the sessions channel
-	// If the socket is closed, signal the channel to close connection
-	buff := make([]byte, bufferSize)
 	for {
+		buff := make([]byte, common.BufferSize)
 		br, err := session.Conn.Read(buff)
+		conf.log.WithError(err).Infof("read %d bytes from conn", br)
+
 		session.Lock.Lock()
-		log.Debugf("read %d bytes from socket, err: %v", br, err)
 		if err != nil {
-			log.Infof("closing session %s; err: %v", session.Id, err)
+			if err == io.EOF {
+				return
+			}
+			conf.log.WithError(err).WithField("session", session.Id).Infof("failed to read from conn")
+
+			// setting Open to false triggers SendData() to
+			// send ShouldClose
 			session.Open = false
 		}
+
+		// write the data to the session buffer, if we have data
 		if br > 0 {
-			session.Buf.Write(buff[:br])
-			if br == len(buff) {
-				newSize := len(buff) * 2
-				log.Infof("increasing buffer size to %d", newSize)
-				buff = make([]byte, newSize)
-			}
+			session.Buf.Write(buff[0:br])
 		}
 		session.Lock.Unlock()
-		if !session.Open {
-			sessions <- session
-			_, _ = common.CloseSession(session.Id)
+
+		sessions <- session
+		if session.Open == false {
 			return
 		}
-		sessions <- session
 	}
 }
 
 func (t *tunnelServer) InitTunnel(stream pb.Tunnel_InitTunnelServer) error {
 	request, err := stream.Recv()
 	if err != nil {
-		log.Fatalf("Failed receiving initial connection from tunnel")
+		return errors.Wrap(err, "failed to read handshake")
 	}
+
```

Comment:
I think it's worthwhile to start a cancellable context in this scope and pass it down to the send and receive functions

Reply:
Agreed, was thinking of going that route!